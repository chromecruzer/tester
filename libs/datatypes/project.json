{
  "name": "datatypes",
  "$schema": "../../node_modules/nx/schemas/project-schema.json",
  "sourceRoot": "libs/datatypes/src",
  "projectType": "library",
  "targets": {
    "lint": {
      "executor": "@nrwl/linter:eslint",
      "outputs": ["{options.outputFile}"],
      "options": {
        "lintFilePatterns": ["libs/datatypes/**/*.ts"]
      }
    },
    "test": {
      "executor": "@nrwl/jest:jest",
      "outputs": ["{workspaceRoot}/coverage/libs/datatypes"],
      "options": {
        "jestConfig": "libs/datatypes/jest.config.ts",
        "passWithNoTests": true
      }
    }
  },
  "tags": []
}
