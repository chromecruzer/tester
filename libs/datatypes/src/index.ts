export * from './datatypes';
export * from './IolRecord';
export * from './CustomerRecord';
export * from './CustomerContactRecord';
export * from './ProductRecord';
export * from './EmployeeRecord';        // added by shyam for testing 
export * from './EmployeeRecordDataMappings'   //
export * from './UserRecord1'  //
export * from './UserRecordDM' // 
export * from './ConsignmentRecord';
export * from './CustomerDataMapping';
export * from './CustomerContactRecord';
export * from './CustomerContactDataMapping';
export * from './IolDataMapping';
export * from './notifications';
export * from './ProductDataMapping';
export * from './SalesRepDataMapping';
export * from './SalesRepRecord';
export * from './SalesMappingSquared';
export * from './SalesMappingRecord';
export * from './SalesJumpRecord';
export * from './UploadHeaders';
export * from './AuditRecord';
export * from './SearchSignature';
export * from './EmailSettings';
//export * from './UserRecord';
import BuildUrl from "./BuildUrl";
import {ExpirationCalculations} from "./ExpirationCalculations";
import ItemMatchStatus from './ItemMatchStatus';
import SearchSignature from "./SearchSignature";
export {BuildUrl, ExpirationCalculations, ItemMatchStatus, SearchSignature};
