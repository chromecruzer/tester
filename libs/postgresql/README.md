# @trac-postgresql

This library was generated with [Nx](https://nx.dev).

## Running unit tests

Run `nx test @trac-postgresql` to execute the unit tests via [Jest](https://jestjs.io).
