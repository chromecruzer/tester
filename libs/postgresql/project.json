{
  "name": "postgresql",
  "$schema": "../../node_modules/nx/schemas/project-schema.json",
  "sourceRoot": "libs/postgresql/src",
  "projectType": "library",
  "targets": {
    "lint": {
      "executor": "@nrwl/linter:eslint",
      "outputs": ["{options.outputFile}"],
      "options": {
        "lintFilePatterns": ["libs/postgresql/**/*.ts"]
      }
    },
    "test": {
      "executor": "@nrwl/jest:jest",
      "outputs": ["{workspaceRoot}/coverage/libs/postgresql"],
      "options": {
        "jestConfig": "libs/postgresql/jest.config.ts",
        "passWithNoTests": true
      }
    }
  },
  "tags": []
}
