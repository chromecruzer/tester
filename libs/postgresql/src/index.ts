export * from './postgresql';
import PostgresConfigFactory, {PostgresConfig} from "./PostgresConfigFactory";
import BulkSqlWithCreate from "./BulkSqlWithCreate";
export *  from './ErrorHandler';
import ExtractIolData from "./ExtractIolData";
import PostgresClient from "./PostgresClient";
import LoadXlsxData from './LoadXlsxData';
import NotificationsServer from "./NotificationsServer";
import SearchTable from "./SearchTable";
import ImportIntoConsignments from "./ImportIntoConsignments";
import ImportIntoJumpTable from "./ImportIntoJumpTable";
import AuditStore from "./AuditStore";
import UserStore from './UserStore';

export {
  AuditStore, PostgresConfig, PostgresConfigFactory, BulkSqlWithCreate, ExtractIolData,
  PostgresClient, LoadXlsxData, NotificationsServer, SearchTable, ImportIntoConsignments, ImportIntoJumpTable,
  UserStore
};
