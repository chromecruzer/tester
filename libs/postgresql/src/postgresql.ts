import * as _ from "lodash";
import {DateTime} from "luxon";
import {PoolClient} from "pg";
import util from "util";
// eslint-disable-next-line @nrwl/nx/enforce-module-boundaries
import {
  customerContactDataMapping,
  customerDataMapping,
  dataTableNames, InputDataType,
  InputDataTypeLabel,
  InputToSqlMapping,
  iolDataMapping,
  matchCustomerContactFields,
  matchCustomerFields,
  matchIolFields,
  matchProductFields,
  matchSalesMappingFields,
  matchSalesRepFields,
  NullableString,
  productDataMapping,
  employeeRecordDm,
  salesMappingDataMapping,
  salesRepDataMapping,
  tracDateFormat, 
  UploadRecord,
  matchEmployeeFields,
  userRecordDm,
  matchUser1Fields
} from "../../datatypes/src";
import path from "path";

const dump = (obj, depth = null) => util.inspect(obj, {depth, colors: true});

export const auditTableNames = {
  notes: 'auditnotes',
  locations: 'auditlocation',
  items: 'audititem',
}



export const searchTableNamesFn = () => [
  dataTableNames.consignments,
  dataTableNames.customers,
  dataTableNames.customerContacts,
  dataTableNames.products,
  dataTableNames.salesReps];

export const sqlDate = (datum: (Date | string)) => {
  // console.log(`datum ${dump(datum)}
  // from JSDate ${dump(DateTime.fromJSDate(datum))}
  // to format ${dump(DateTime.fromJSDate(datum).toFormat(tracDateFormat))}`)
  const date = _.isDate(datum) ? datum : new Date(datum);
  return DateTime.fromJSDate(date).toFormat(tracDateFormat)
}

export function datumToSql(datum: (string | number | Date | boolean), datatype: InputDataTypeLabel) {
  if (_.isNull(datum) || _.isUndefined(datum) || datum === '') {
    return 'NULL';
  }
  let str;
  switch (datatype) {
    case 'string':
      str = String(datum);
      if (str.length === 0) {
        return null;
      }
      str = str.replace(/'/g, "''");
      str = str.replace(/\n/g, "\\n");
      return `'${str}'`;
    case 'date':
      // console.log(`date ${datum} converted ${sqlDate(datum as unknown as Date)}`)
      return `'${sqlDate(datum as Date)}'`
    case 'boolean':
      // console.log(`boolean ${datum} as ${Boolean(datum)}`)
      return Boolean(datum);
    case 'number':
      return datum;
  }
}

export function appendUuids(sql: string, uuids: string[]) {
  let skipped1st = false;
  let result = sql;
  uuids.forEach(id => {
    result += `${skipped1st ? ',' : ''}    \n'${id}'`;
    skipped1st = true;
  });
  result += ');';
  return result;
}

export const checkModified = modified => {
  // console.log(`Checking ${dump(modified)}`);
  if (modified.originalUuid !== modified.new.uuid) {
    console.error(`UUID mismatch found ${dump(modified)}`);
  }
}

export function appendBulkData(sql, bulkData: UploadRecord[],
                               columnMapping: InputToSqlMapping[],
                               createUuid = true,
                               hasUuid = false) {
  let skipped1stRow = false;
  let result = `${sql}\n`;
  bulkData.forEach((record) => {
    // console.log(`append bulk data ${dump(record)}`)
    const dataRow = [];
    if (createUuid) {
      dataRow.push('gen_random_uuid()')
    }
    if(hasUuid) {
      dataRow.push(datumToSql(record['uuid'], 'string'));
    }
    columnMapping.forEach((colMap) => {
      let datum = datumToSql(record[colMap.sqlLabel], colMap.dataType);
      if(!_.isUndefined(colMap.default) && datum === null) {
        datum = colMap.default;
      }
      dataRow.push(datum)
    });
    result += `${skipped1stRow ? ',': ''}\n(${_.join(dataRow, ', ')})`;
    skipped1stRow = true;
  });
  return result;
}

export const queryRecords = async (client: PoolClient, recordSql: string, idField: string,
                                   {
                                     uuid = null as NullableString,
                                     id = null as NullableString,
                                     ids = [] as string[],
                                   }) => {
  let sql = recordSql;
  const check = (_.isString(uuid) ? 1 : 0) + (_.isString(id) ? 1 : 0) + (_.isEmpty(ids) ? 0 : 1);
  if (check > 1) {
    throw new Error(`cannot set multiple parameters ${uuid}, ${id}, ${ids.length}`);
  }
  if (uuid) {
    sql += ` WHERE uuid='${uuid}'`;
  }
  if (id) {
    sql += ` WHERE ${idField}='${id}'`;
  }
  if (!_.isEmpty(ids)) {
    sql += ` WHERE ${idField} IN (${_.join(_.map(ids, i => `'${i}'`), ', ')})`;
  }
  sql += ';';
  // console.log(`query records sql '${sql}'`);
  return client.query(sql)
    .then(data => data.rows)
    .catch(err => {
    throw new Error(`Mistake in '${sql}' caused ${err.message}`);
  });
}

interface TransformerKeys {
  matchFields: (string | string[]);
  matchFn: (a: any, b: any) => boolean,
  mappings: InputToSqlMapping[];
  compareFieldList?: string[],
}

export const recordFromArray = (row: InputDataType[], mapping: InputToSqlMapping[]) => {
  return _.reduce(mapping, (accum, m,i) => {
    const r = row[i] || null;
    let value = _.isFunction(m.nullify) ? m.nullify(r) : r;
    if(_.has(m, 'default')) { //need to us _.has here because the default may be a boolean false
      // console.log(`checking default for ${dump(m)} with value ${value}`)
      value = value || m.default;
    }
    accum[m.sqlLabel] = value;
    return accum;
  }, {} as UploadRecord);
}
export const transformHelpers: { [id: string]: TransformerKeys } = {
  iol: {
    matchFields: 'lot',
    matchFn: matchIolFields,
    mappings: iolDataMapping,
  },
  customers: {
    matchFields: 'customer_code',
    matchFn: matchCustomerFields,
    mappings: customerDataMapping,
  },
  customer_contacts: {
    matchFields: ['full_name', 'ship_to_id'],
    matchFn: matchCustomerContactFields,
    mappings: customerContactDataMapping,
  },
  products: {
    matchFields: 'product_id',
    matchFn: matchProductFields,
    mappings: productDataMapping,
  },
  employee: {
     matchFields: 'emp_id',
     matchFn: matchEmployeeFields,          // added by shyam for testing
     mappings: employeeRecordDm,
  },

  users: {
    matchFields: 'username',
    matchFn:  matchUser1Fields,              // added by shyam for testing
    mappings: userRecordDm
  },
  salesmappings: {
    matchFields: 'customer_ship_to',
    matchFn: matchSalesMappingFields,
    mappings: salesMappingDataMapping,
    compareFieldList: ['customer_ship_to', 'postal', 'zipcode', 'primary_territory', 'primary_name', 'cos_territory',
      'cos_name', 'kae_territory', 'kae_name'],
  },
  salesreps: {
    matchFields: 'salesrep_id',
    matchFn: matchSalesRepFields,
    mappings: salesRepDataMapping,
  },
}
export const matchFieldFn = (fields, record) => {
  return _.isArray(fields) ? _.join(_.map(fields, f => record[f]), '|') : record[fields]
}
export const deepEqualFn = (c, m, compareFieldList = null) => {
  if (_.isArray(compareFieldList)) {
    return _.reduce(compareFieldList,
      (boolAccum, cf) => {
        if (!boolAccum) {
          return boolAccum;
        }
        // console.log(`comparing isequal field ${cf}, ${candidate[cf]} to ${m[cf]}`);
        return _.isEqual(c[cf], m[cf]);
      }, true);
  }
  return _.isEqual(m, c);
}
export const docxTemplateFn = (apiConfig, templateType) => path.join(apiConfig.templatesPath,
  (templateType === 'Missing' ? 'missingTemplate.docx' : 'expiredTemplate.docx'))
export const htmlTemplateFn = (apiConfig, templateType) => path.join(apiConfig.templatesPath,
  (templateType === 'Missing' ? 'missingTemplate.html' : 'expiredTemplate.html'))

