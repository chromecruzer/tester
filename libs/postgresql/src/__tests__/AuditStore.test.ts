import {AuditStore} from "@trac/postgresql";
import * as Datatypes from "@trac/datatypes"
import {DateTime} from "luxon";
import {tracDateFormat} from "@trac/datatypes";
describe('AuditStore', () => {
  let store, mockClient;
  beforeEach(() => {
    store = new AuditStore('name of the schema');
    mockClient = {
      query: jest.fn().mockReturnValue(Promise.resolve())
    };
    jest.spyOn(DateTime, 'now').mockImplementation(() => DateTime.fromFormat('07/20/2022',tracDateFormat))
  });
  describe('storeNotes', () => {
    let oneNote, multipleNotes;
    beforeEach(() => {
      oneNote = {
        date_created: '07/04/2022',
        author: 'Dickens',
        annotate_type: 'Item',
        annotated_uuid: 'uuid-of-the-note-target',
        audit_uuid: 'uuid-of-the-audit',
        content: `note that show the content
        with an embedded`
      }
      multipleNotes = [
        {
          date_created: '07/04/2022',
          author: 'Dickens',
          annotate_type: 'Item',
          annotated_uuid: 'uuid-of-the-note-target',
          audit_uuid: 'uuid-of-the-audit',
          content: `note that show the content
        with an embedded return`
        },
        {
          date_created: '07/11/2022',
          author: 'Shakespeare',
          annotate_type: 'Audit',
          annotated_uuid: 'uuid-of-the-note-target2',
          audit_uuid: 'uuid-of-the-audit2',
          content: `note that show the content`
        },
        {
          date_created: '07/04/2022',
          author: 'Verne',
          annotate_type: 'Match',
          annotated_uuid: 'uuid-of-the-note-target3',
          audit_uuid: 'uuid-of-the-audit3',
          content: `note that show the content
        with an embedded
        returns`
        },
      ];
    });
    it('should store one note correctly', () => {
      return store.storeNotes([oneNote], mockClient)
        .then(() => {
          expect(mockClient.query.mock.calls[0][0]).toMatchSnapshot();
        });
    });
    it('should store multiple notes correctly', () => {
      store.storeNotes(multipleNotes, mockClient)
        .then(() => {
          expect(mockClient.query.mock.calls[0][0]).toMatchSnapshot();
        });
    });
  });
  describe('updateItems', () => {
    it('should update one audit item correctly', () => {
      store.updateItems(['uuid1'], 'Billed', 'Nutall', mockClient, null, null, warehouse)
        .then(() => {
          expect(mockClient.query.mock.calls[0][0]).toMatchSnapshot();
        });
    });
    it('should update multiple audit items correctly', () => {
      store.updateItems(['uuid1', 'uuid2', 'uuid3'], 'Billed', 'Nutall', mockClient, null, null, warehouse)
        .then(() => {
          expect(mockClient.query.mock.calls[0][0]).toMatchSnapshot();
        });
    });
    it('should update multiple audit items with consignment location correctly', () => {
      store.updateItems(['uuid1', 'uuid2', 'uuid3'], 'Found In Another Location', 'Nutall', mockClient, 'uuid of the consignment', 'location of the consignment', warehouse)
        .then(() => {
          expect(mockClient.query.mock.calls[0][0]).toMatchSnapshot();
        });
    });
  });
});
