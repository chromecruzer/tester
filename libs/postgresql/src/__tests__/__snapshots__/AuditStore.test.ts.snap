// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuditStore storeNotes should store multiple notes correctly 1`] = `
"INSERT INTO name of the schema.auditnotes VALUES (gen_random_uuid(),
    '07/04/2022',
    'Dickens',
    'Item',
    'uuid-of-the-note-target',
    'uuid-of-the-audit',
    'note that show the content\\\\n        with an embedded return'),
(gen_random_uuid(),
    '07/11/2022',
    'Shakespeare',
    'Audit',
    'uuid-of-the-note-target2',
    'uuid-of-the-audit2',
    'note that show the content'),
(gen_random_uuid(),
    '07/04/2022',
    'Verne',
    'Match',
    'uuid-of-the-note-target3',
    'uuid-of-the-audit3',
    'note that show the content\\\\n        with an embedded\\\\n        returns');"
`;

exports[`AuditStore storeNotes should store one note correctly 1`] = `
"INSERT INTO name of the schema.auditnotes VALUES (gen_random_uuid(),
    '07/04/2022',
    'Dickens',
    'Item',
    'uuid-of-the-note-target',
    'uuid-of-the-audit',
    'note that show the content\\\\n        with an embedded');"
`;

exports[`AuditStore updateItems should update multiple audit items correctly 1`] = `"UPDATE name of the schema.audititem SET audit_match = 'Billed' WHERE uuid IN ('uuid1','uuid2','uuid3');"`;

exports[`AuditStore updateItems should update multiple audit items with consignment location correctly 1`] = `
"UPDATE name of the schema.audititem SET audit_match = 'Found In Another Location', consignment_uuid = 'uuid of the consignment', consignment_location = 'location of the consignment',
      confirmed_date = '07/20/2022', confirmed_by = 'Nutall'  WHERE uuid IN ('uuid1','uuid2','uuid3');"
`;

exports[`AuditStore updateItems should update one audit item correctly 1`] = `"UPDATE name of the schema.audititem SET audit_match = 'Billed' WHERE uuid='uuid1';"`;
